pulp_2to3_migration/app/plugin/rpm/migrator.py

```diff
@@ -126,6 +158,108 @@ def pipeline_stages(self):
         return pipeline
 
 
+class RpmContentMigrationFirstStage(ContentMigrationFirstStage):
+    """
+    The first stage of a Rpm content migration pipeline.
+    """
+
+    async def migrate_to_pulp3(self, batch, pb=None):
+        """
+        A default implementation of DeclarativeContent creation for migrating content to Pulp 3.
+
+        Plugin writers might want to override this method if it doesn't satisfy their needs as is.
+
+        In this implementation there is an assumption that each content has one artifact.
+
+        Args:
+            batch: A batch of Pulp2Content objects to migrate to Pulp 3
+        """
+        @functools.lru_cache(maxsize=20)
+        def get_remote_by_importer_id(importer_id):
+            """
+            Args:
+                importer_id(str): Id of an importer in Pulp 2
+
+            Returns:
+                remote(pulpcore.app.models.Remote): A corresponding remote in Pulp 3
+
+            """
+            try:
+                pulp2importer = Pulp2Importer.objects.get(pulp2_object_id=importer_id)
+            except ObjectDoesNotExist:
+                return
+            return pulp2importer.pulp3_remote
+
+        for pulp_2to3_detail_content in batch:
+            pulp2content = pulp_2to3_detail_content.pulp2content
+
+            if getattr(pulp2content, 'downloaded'):
+
+                # get all Lazy Catalog Entries (LCEs) for this content
+                pulp2lazycatalog = Pulp2LazyCatalog.objects.filter(
+                    pulp2_unit_id=pulp2content.pulp2_id)
+
+                if not pulp2content.downloaded and not pulp2lazycatalog:
+                    _logger.warn(_('On_demand content cannot be migrated without an entry in the '
+                                   'lazy catalog, pulp2 unit_id: '
+                                   '{}'.format(pulp2content.pulp2_id)))
+                    continue
+
+            pulp3content = await pulp_2to3_detail_content.create_pulp3_content()
+            future_relations = {'pulp2content': pulp2content}
+
+            comps_types = ('package_langpacks', 'package_group', 'package_category',
+                           'package_environment')
+            if pulp_2to3_detail_content.pulp2_type in comps_types:
+                # dc without artifact
+                dc = DeclarativeContent(content=pulp3content)
+                dc.extra_data = future_relations
+                await self.put(dc)
```

Comment:
I had to do similar thing for errata and I wonder if we can generalize that and not copy-paste-adjust the migrate_to_pulp3.
E.g. I could imagine having different methods for no-artifact, one-artifact, multiple-artifacts cases.
And in the migrator we can configure which content is which.

WDYT?

Reply:
i think this is a good idea to consider later. I think we might still need to copy migrate_to_pulp3 due to content relation.
My plan was to come back to this piece of code after relating content to see if this is possible to generalize.

Reply:
We should just be careful that the branches don't get out of sync any more than they have to.  I would recommend marking any changed sections with comments where they begin and end, so that we know exactly what has changed, and what has not.

Reply:
It would also make reviewing easier - less manual comparing the two.